src/components/TodoItem/TodoItem.tsx

```diff
@@ -0,0 +1,129 @@
+import React, { memo, useState } from 'react';
+import classNames from 'classnames';
+import { Todo } from '../../types/Todo';
+import { TodoPatch } from '../../types/TodoPatch';
+
+type Props = {
+  todo: Todo,
+  onDelete: (todoId: number) => void,
+  onUpdateTodo?: (todoId: number, data: TodoPatch) => void,
+  isLoading: boolean
+};
+
+export const TodoItem: React.FC<Props> = memo(
+  ({
+    todo,
+    onDelete,
+    onUpdateTodo = () => { },
+    isLoading,
+  }) => {
+    const {
+      id,
+      title,
+      completed,
+    } = todo;
+
+    const [isEditing, setIsEditing] = useState(false);
+    const [newTitle, setNewTitle] = useState(title);
+
+    const handleDeleteTodo = () => onDelete(id);
+
+    const handleToggleTodo = () => onUpdateTodo(id, { completed: !completed });
+
+    const handleToggleEditing = () => setIsEditing(state => !state);
+
+    const handleChangeNewTitle = (
+      event: React.ChangeEvent<HTMLInputElement>,
+    ) => {
+      setNewTitle(event.target.value);
+    };
+
+    const handleSubmit = (event: React.FormEvent) => {
+      event.preventDefault();
+
+      if (!newTitle.trim().length) {
+        onDelete(id);
+      } else if (title !== newTitle) {
+        setNewTitle(prevTitle => prevTitle.trim());
+        onUpdateTodo(id, { title: newTitle });
+      }
+
+      handleToggleEditing();
+    };
+
+    const handleBlur = (event: React.FormEvent) => {
+      handleSubmit(event);
+    };
+
+    const handleCancelEditing = (
+      event: React.KeyboardEvent<HTMLInputElement>,
+    ) => {
+      if (event.key === 'Escape') {
```

Comment:
It would be cool not to use magic values and put such strings into constants and reuse them:
```
const ESCAPE = 'Escape';
```
This will increase the readability of the code and save us from possible errors.
This applies to development in general.